### DISCUSSION TITLE
Extension point to wire in domain data

### DISCUSSION THREAD

**Discussion by dil-abertok** [bchavez/Bogus#523] [Q&A]
Hi there,

When I am generating generic data (people, addresses, etc.), Bogus has the really lovely fluent syntax with the `DataSets` as properties. Now, I usually create my own methods to create bogus instances of simple domain concepts, say, my `id`s are all 8 digit hexadecimal, so I have a `CreateId(Faker faker)` method, that uses the supplied faker to generate just that without me having to express that same rule everywhere I see an `Id` property.
In order to wire that in, I either have a `DomainFaker` static class, and I do something like `.RuleFor(x => x.Id, f => DomainFaker.Id(f))`, or I create an instance class and extension method on `Faker` that returns me that instance class (passing along the faker instance), something like `.RuleFor(x => x.Id, f => f.Domain().Id()`. However, the extra pair of brackets after `Domain` sort of bugs me as idiosyncratic as none of the actual `DataSet`s needs that.
Would it be possible to extend `Faker` with an "empty" `Domain` (partial?) class that people could use to add their extension methods to? ("Empty" as in it would probably need a property that references back the original `Faker` in order to have that available in the extension methods. Making it partial would probably allow for better extensibility.) That way we could do something like `.RuleFor(x => x.Id, f => f.Domain.Id()`.

Cheers, and thanks for the awesome package!

**Comment by bchavez** [comment]
Hi. Thank you for the post.

You'll have to explain the "`partial`" approach more. AFAIK, the C# `partial` keyword won't work in this case because:
* https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/partial-classes-and-methods
>All partial-type definitions meant to be parts of the same type must be defined in the same assembly and the same module (.exe or .dll file). Partial definitions cannot span multiple modules.

This means that `public partial class Faker { myStuff }` would only work for source code inside and defined in `Bogus.dll` at compile time; `partial` class definitions can't span multiple DLLs or EXE projects.

I agree, the two extra `()` in `f => f.Domain().MyMethod()` is a bit tacky language and ergonomic-wise, **_but it is the way_**.

I think, ideally, a C# extension property would work best here; but C# extension properties aren't a thing yet. 😭
* https://stackoverflow.com/questions/619033/does-c-sharp-have-extension-properties

So, the next best thing, alternatively, might be to add a dedicated `.RuleFor` extension method to `Faker<T>` for domain-specific data:
```csharp
void Main()
{
   var faker = new Faker<MyDomainObject>()
      .RuleFor(x => x.SpecialValue, fd => fd.Domain.SpecialString());

   faker.Generate().Dump();
}

public class MyDomainObject
{
   public string SpecialValue;
}

public static class ExtensionsForDomainFakerT
{
   public static Faker<T> RuleFor<T, TProperty>(this Faker<T> fakerT, Expression<Func<T, TProperty>> property, Func<FakerWithDomain, TProperty> setter)
      where T : class
   {
      return fakerT
         .RuleFor(property, f =>
         {
            var domainFaker = new FakerWithDomain(f);
            return setter(domainFaker);
         });
   }
}

public class FakerWithDomain : Faker
{
   public FakerWithDomain(Faker existing) : base()
   {
      this.Domain = this.Notifier.Flow(new DomainDataSet());
      this.Random = existing.Random;
   }
   public DomainDataSet Domain { get; init; }
}

public class DomainDataSet : Bogus.DataSet
{
   public string SpecialString()
   {
      var value = this.Random.Replace("???###");
      return value;
   }
}
```
![image](https://github.com/bchavez/Bogus/assets/478118/c0ee31a0-307a-4b91-93c1-5aa3e333819d)

The above code seems to work on the .NET 8 SDK. It's a bit verbose to get the "`Faker<T>`" stuff wired up, but it seems to work.

If not, try changing `.RuleFor() -> .RuleForDomain()` in `ExtensionsForDomainFakerT` -- in this case, you'll still have to deal with some kind of uniqueness for the C# compiler to resolve the ambiguous method call. Such is life. 😞 

There are probably a few ways to hack it with the DLR (dynamic language runtime and the `dynamic` keyword), but then you start running into possible perf issues and maybe losing IntelliSense.

Feel free to carry on the discussion after I close the thread.

Hope that helps,
Brian Chavez
